### PR TITLE
[ops] feat: add the TileLang teacher-distribution kernel for the DeepSeek-V4 indexer

### PR DESCRIPTION
**Pull request by FoolPlayer**
### What does this PR do?

Adds the TileLang kernel that computes the DeepSeek-V4 Lightning Indexer's **teacher
distribution**, and lets the existing sparse attention forward hand back the log-sum-exp it
already computes.

Kernel and plumbing only. The KL objective that consumes this arrives in a later PR, so nothing
here changes training behaviour.

### Checklist Before Starting

- Search for relative PRs/issues and link here: builds on #1012 (sparse MLA kernel time),
  already merged.
- PR title follows `[{modules}] {type}: {description}` format

### Test

- `pytest tests/ops/test_deepseek_v4_indexer_loss.py` — 19 passed. The suite includes an
  independent PyTorch reference for the teacher distribution, `NI > 1` coverage, and the
  zero-row contract.
- `make quality` — clean.
- `tests/ops` already runs as a whole directory in `gpu_unit_tests.yml`, so no workflow change
  is needed.

### API and Usage Example

```python
from veomni.ops.kernels.deepseek_v4 import sparse_attn_tilelang, sparse_mqa_target_fwd

out, lse = sparse_attn_tilelang(q, kv, sink, topk_idxs, sm_scale, return_lse=True)
target = sparse_mqa_target_fwd(q, kv, topk_idxs, lse, sm_scale)
```

`return_lse` defaults to `False`, so every existing call site is unaffected.

### Design & Code Changes

**`sparse_mqa_target_fwd`** produces the per-candidate attention probabilities the indexer is
distilled against, renormalized with the LSE from the attention forward rather than recomputing
a softmax. Its guards `raise` rather than `assert` so they survive `python -O`.

**Optional LSE.** The forward kernel already writes the LSE; returning it costs nothing. It is
marked non-differentiable in the autograd `Function`: the LSE only ever feeds the detached
indexer teacher, and leaving it differentiable would open a second gradient route back into
`q`/`kv`/`sink` and break the decoupling the indexer objective depends on.

### Checklist Before Submitting

- Applied pre-commit checks
- Added/updated documentation
- Added tests to CI workflow (covered by the existing `tests/ops` directory run)

## Summary by CodeRabbit

* **New Features**
  * Sparse attention can optionally return log-sum-exp (LSE) values alongside attention outputs.
  * Added target attention support for compressed sparse indices on compatible CUDA hardware.
  * Improved handling of invalid or padded sparse indices and zero-mass rows.
* **Bug Fixes**
  * Added validation for unsupported head counts, data types, shapes, and configurations.
  * Ensured returned LSE values are not differentiated during backpropagation while attention outputs remain trainable.
* **Tests**
  * Added comprehensive coverage for normalization, masking, warnings, gradients, and kernel correctness.

### CODE REVIEW

**Comment by FoolPlayer** [comment]
Both findings addressed in e900ee621.

**`S_kv > 0` guard.** Valid — the analysis is right that the candidate mask cannot prevent the gather, since it only zeroes the score afterwards. Added as a raise-style guard to match the rest of the interface, so `python -O` cannot strip it, plus a regression test for the empty-kv case.

**Hard-coded `"cuda"`.** Already fixed before this review landed; the tests use `DEVICE = get_device_type()` and `device-api-check` is green.

**Comment by FoolPlayer** [comment]
@coderabbitai review

**Review by Coach257 — APPROVED** [review]
